File: 2025/340. Longest Substring with At Most K Distinct Characters/Solution.java
class Solution {
    public int lengthOfLongestSubstringKDistinct(String s, int k) {
        Map<Character, Integer> map = new HashMap<>();
        char[] arr = s.toCharArray();
        int i = 0, max = 0;
        for (int j = 0; j < arr.length; j++) {
            map.put(arr[j], map.getOrDefault(arr[j], 0) + 1);
            while (i <= j && map.size() > k) {
                if (map.get(arr[i]) == 1)
                    map.remove(arr[i]);
                else 
                    map.put(arr[i], map.get(arr[i]) - 1);
                i++;
            }
            max = Math.max(j - i + 1, max);
        }
        return max;
    }
}

File: 2025/340. Longest Substring with At Most K Distinct Characters/solution.py
class Solution:
    def lengthOfLongestSubstringKDistinct(self, s: str, k: int) -> int:
        sliding_win = defaultdict(int)
        l, res = 0, 0
        for r in range(len(s)):
            ch = s[r]
            sliding_win[ch] += 1
            while len(sliding_win) > k:
                sliding_win[s[l]] -= 1
                if sliding_win[s[l]] == 0:
                    del sliding_win[s[l]]
                l += 1
            res = max(r - l + 1, res)
        return res